nft-client/src/main.rs

```diff
@@ -0,0 +1,635 @@
+                tx.effects
+            {
+                if data.address == contract {
+                    return data.events;
+                }
+            }
+        }
+    }
+    Vec::new()
+}
+
+/// Convert an account address from `types` to an account address from
+/// `concordium_contracts_common`.
+fn convert_account_address(
+    account: &id::types::AccountAddress,
+) -> concordium_contracts_common::AccountAddress {
+    let mut address_bytes = [0u8; 32];
+    address_bytes.copy_from_slice(account.as_ref());
+    concordium_contracts_common::AccountAddress(address_bytes)
+}
+
+/// Convert a contract address from `types` to a contract address from
+/// `concordium_contracts_common`.
+fn convert_contract_address(
+    contract: &types::ContractAddress,
+) -> concordium_contracts_common::ContractAddress {
+    concordium_contracts_common::ContractAddress {
+        index:    contract.index.index,
+        subindex: contract.subindex.sub_index,
+    }
+}
+
+/// Convert an address from `types` to an address from
+/// `concordium_contracts_common`.
+fn convert_address(address: types::Address) -> concordium_contracts_common::Address {
+    match address {
+        types::Address::Account(addr) => {
+            concordium_contracts_common::Address::Account(convert_account_address(&addr))
+        }
+        types::Address::Contract(addr) => {
+            concordium_contracts_common::Address::Contract(convert_contract_address(&addr))
+        }
+    }
+}
+
+/// Print the NFT contract state in a human readable format.
+fn pretty_print_contract_state(state: &NFTContractState) {
+    for (owner, address_state) in &state.0 {
+        if address_state.owned_tokens.is_empty() && address_state.operators.is_empty() {
+            continue;
+        }
+        match owner {
+            concordium_contracts_common::Address::Account(addr) => println!("\n\n{}", addr),
+            concordium_contracts_common::Address::Contract(addr) => println!("\n\n{}", addr),
+        }
+        if !address_state.owned_tokens.is_empty() {
+            println!("  Owned Token IDs");
+            for token in &address_state.owned_tokens {
+                println!("    - {}", token);
+            }
+        }
+        if !address_state.operators.is_empty() {
+            println!("  Operators");
+            for operator in &address_state.operators {
+                println!("    - {:?}", operator);
+            }
+        }
+    }
+}
+
+/// Build and send a transaction from a payload
+async fn send_transaction(
+    client: &mut Client,
+    account_data: &AccountData,
+    payload: transactions::Payload,
+) -> anyhow::Result<types::hashes::TransactionHash> {
+    let next_account_nonce = client.get_next_account_nonce(&account_data.address).await?;
+    let expiry =
+        common::types::TransactionTime::from_seconds((chrono::Utc::now().timestamp() + 300) as u64);
+    let energy = transactions::send::GivenEnergy::Add(3000.into());
```

Comment:
Where does `3000` come from?

Reply:
I have replaced this with an estimate measure from experiments.

A more precise measurement would probably require fetching the size of the contract state.